---
title: "Gene expression statistical analysis"
author: "Davide Risso"
date: "24/01/2018"
output:
  beamer_presentation:
    includes:
      in_header: template.tex
    toc: yes
    keep_tex: no
---

# Introduction

```{r options, include=FALSE}
knitr::opts_chunk$set(cache=TRUE, error=FALSE, message=FALSE, warning=FALSE,
                      echo=FALSE, results="markup")
library(RColorBrewer)
library(EDASeq)
library(DESeq2)
library(RUVSeq)
library(edgeR)
```


## What we will cover

We will cover differential expression analysis of RNA-seq data in _R/Bioconductor_.

I'm assuming familiarity with the R programming language for the examples.

We will start from a _matrix of gene-level read counts_.

We will cover the two most popular packages, `DESeq2` and `edgeR`.

I will also show you how to deal with unwanted variation using the `RUVSeq` package.

## What we will not cover

Only in Bioconductor, there are 150 packages for QC/EDA, 69 for normalization, and 241 for differential expression! 

Hence, this is not a comprehensive account on how to perform these steps, but rather an _introduction to the statistical methods_ behind some of them.

## Useful links

- These slides: [https://github.com/drisso/canazei](github.com/drisso/canazei)
- Example dataset: [https://github.com/drisso/peixoto2015_tutorial](github.com/drisso/peixoto2015_tutorial)

- The edgeR user guide [https://bioconductor.org/packages/edgeR](bioconductor.org/packages/edgeR)
- The DESeq2 vignette  [https://bioconductor.org/packages/DESeq2](bioconductor.org/packages/DESeq2)
- The F1000 Research Bioconductor gateway [https://f1000research.com/gateways/bioconductor](f1000research.com/gateways/bioconductor)
- Bioconductor support forum [https://support.bioconductor.org](support.bioconductor.org)

## Gene-level read counts

\scriptsize
```{r, echo=FALSE, results='markup'}
data_dir <- "~/git/peixoto2015_tutorial/Peixoto_Input_for_Additional_file_1/"
fc <- read.table(paste0(data_dir, "Peixoto_CC_FC_RT.txt"), row.names=1, header=TRUE)
negControls <- read.table(paste0(data_dir, "Peixoto_NegativeControls.txt"), sep='\t', header=TRUE, as.is=TRUE)
positive <- read.table(paste0(data_dir, "Peixoto_positive_controls.txt"), as.is=TRUE, sep='\t', header=TRUE)

x <- as.factor(rep(c("CC", "FC", "RT"), each=5))
names(x) <- colnames(fc)

filter <- apply(fc, 1, function(x) length(x[which(x>10)])>5)
filtered <- as.matrix(fc)[filter,]
head(filtered)
```

## An example dataset

![](FCdesign.png)

## An example dataset

- C57BL/6J adult male mice (2 months of age). 
- Five animals per group: fear conditioning (FC), memory retrieval (RT), and controls (CC).
- Illumina 100bp paired-end reads mapped to the mouse genome (mm9) using GMAP/GSNAP.
- Ensembl (release 65) gene counts obtained using HTSeq.

# Exploratory Data Analysis

```{r}
negCon <- intersect(negControls[,2], rownames(filtered))
FCup <- intersect(positive[positive[,3]=="UP",1], rownames(filtered))
FCdown <- intersect(positive[positive[,3]=="DOWN",1], rownames(filtered))
RTup <- intersect(positive[positive[,4]=="UP",1], rownames(filtered))
RTdown <- intersect(positive[positive[,4]=="DOWN",1], rownames(filtered))

colors <- brewer.pal(9, "Set1")
colLib <- colors[x]
```

## Exploratory Data Analysis

## Relative Log Expression (RLE)

## Principal Component Analysis

# Normalization

## Accounting for sequencing depth

## Upper-quartile

## Trimmed Mean of M-values (TMM)

## The DESeq normalization

# Differential Expression

## Statistical tests of hypothesis

## The t-test and linear models

## The Poisson Model

When statisticians see counts, they immediately think about Simeon Poisson.

\centering
\includegraphics[width=.7\linewidth]{Simeon_Poisson}

## The Poisson Model

The Poisson distribution naturally arises from binomial calculations, with a large number of trials and a small probability.

It has a rather stringent assumption: **the variance is equal to the mean**!

$$
Var(Y_{ij}) = \mu_{ij}
$$

In real datasets the variance is greater than the mean, a condition known as **overdispersion**.

## A real example

```{r, echo=FALSE, message=FALSE, warning=FALSE}
y <- DGEList(counts = filtered, group = x)
y <- calcNormFactors(y)
design <- model.matrix(~x)
y <- estimateDisp(y, design)
meanVarPlot <- plotMeanVar(y, 
                           show.raw.vars=TRUE, 
                           show.tagwise.vars=FALSE,
                           show.binned.common.disp.vars=FALSE,
                           show.ave.raw.vars=FALSE, 
                           NBline = TRUE , nbins = 100,
                           pch = 16, 
                           xlab ="Mean Expression (Log10 Scale)", 
                           ylab = "Variance (Log10 Scale)" , 
                           main = "Mean-Variance Plot" )
```

## The Negative Binomial Model

A generalization of the Poisson model is the negative binomial, that assumes that the variance is a quadratic function of the mean.

$$
Var(Y_{ij}) = \mu_{ij} + \phi_j \mu_{ij}^2
$$
where $\phi$ is called the **dispersion parameter**.

Both `edgeR` and `DESeq2` assume that the data is distributed as a negative binomial.

## Differential expression in R: `DESeq2`

## Diagnostics: the p-value histogram

## Visualizing the results

# Batch Effects

## Accounting for known batch effects

## Accounting for unknown batch effects

# Transcript-level analysis

# Single-cell RNA-seq


